testutil/mockconn.go

```diff
@@ -0,0 +1,155 @@
+package testutil
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+var (
+	connIsClosed = &mutexBool{
+		mutex: &sync.Mutex{},
+		state: false,
+	}
+	connIsReset = &mutexBool{
+		mutex: &sync.Mutex{},
+		state: false,
+	}
+)
+
+type mutexBool struct {
+	mutex *sync.Mutex
+	state bool
+}
+
+// Conn facilitates testing by providing two connected ReadWriteClosers
+// each of which can be used in place of a net.Conn
+type Conn struct {
+	Server *End
+	Client *End
+}
+
+// Close closes server/client pipes
+func (c *Conn) Close() error {
+	if err := c.Server.Close(); err != nil {
+		return err
+	}
+	if err := c.Client.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewConn returns a new testutil.Conn
+func NewConn() *Conn {
+	// A connection consists of two pipes:
```

Comment:
Since a net.Pipe is a ReadWriter, it represents a duplex channel. I'm not sure you need two of these, which means you probably don't need a lot of the other code you have around these.